include/sparrow/arrow_array_schema_proxy.hpp

```diff
@@ -96,99 +135,170 @@ namespace sparrow
-        SPARROW_ASSERT_TRUE((this->array().dictionary == nullptr) == (this->schema().dictionary == nullptr));
+    inline void arrow_proxy::validate_array_and_schema() const
+    {
+        SPARROW_ASSERT_TRUE(m_array_ref.release != nullptr);
+        SPARROW_ASSERT_TRUE(m_schema_ref.release != nullptr);
+        SPARROW_ASSERT_TRUE(m_array_ref.n_children == m_schema_ref.n_children);
+        SPARROW_ASSERT_TRUE((m_array_ref.dictionary == nullptr) == (m_schema_ref.dictionary == nullptr));
 
-        const data_type data_type = format_to_data_type(this->schema().format);
-        if (!validate_format_with_arrow_array(data_type, array()))
+        const enum data_type data_type = format_to_data_type(m_schema_ref.format);
+        if (!validate_format_with_arrow_array(data_type, m_array_ref))
         {
-            // TODO: Replace with a more specific exception
-            throw std::runtime_error("Invalid ArrowArray format");
+            throw arrow_proxy_exception("Invalid ArrowArray format");
         }
     }
 
-    inline
-    arrow_proxy::arrow_proxy(ArrowArray&& array, ArrowSchema&& schema)
+    template <typename AA, typename AS>
+    arrow_proxy::arrow_proxy(AA&& array, AS&& schema, bool)
         : m_array(std::move(array))
+        , m_array_ref(m_array.index() == 0 ? *std::get<0>(m_array) : std::get<1>(m_array))
         , m_schema(std::move(schema))
+        , m_schema_ref(m_schema.index() == 0 ? *std::get<0>(m_schema) : std::get<1>(m_schema))
     {
+        if constexpr (std::is_rvalue_reference_v<AA&&>)
+        {
+            array = {};
+        }
+        else if constexpr (std::is_pointer_v<std::remove_cvref_t<AA>>)
+        {
+            SPARROW_ASSERT_TRUE(array != nullptr);
+        }
+
+        if constexpr (std::is_rvalue_reference_v<AS&&>)
+        {
+            schema = {};
+        }
+        else if constexpr (std::is_pointer_v<std::remove_cvref_t<AS>>)
+        {
+            SPARROW_ASSERT_TRUE(schema != nullptr);
+        }
+
         validate_array_and_schema();
-        initialize_children();
         initialize_buffers();
     }
 
-    inline
-    arrow_proxy::arrow_proxy(ArrowArray&& array, ArrowSchema* schema)
-        : m_array(std::move(array))
-        , m_schema(schema)
+    inline arrow_proxy::arrow_proxy(ArrowArray&& array, ArrowSchema&& schema)
+        : arrow_proxy(std::move(array), std::move(schema), true)
+    {
+    }
+
+    inline arrow_proxy::arrow_proxy(ArrowArray&& array, ArrowSchema* schema)
+        : arrow_proxy(std::move(array), schema, true)
+    {
+    }
+
+    inline arrow_proxy::arrow_proxy(ArrowArray* array, ArrowSchema* schema)
+        : arrow_proxy(array, schema, true)
+    {
+    }
+
+    inline arrow_proxy::arrow_proxy(const arrow_proxy& other)
+        : m_array(&other.m_array_ref)
+        , m_array_ref(*std::get<0>(m_array))
+        , m_schema(&other.m_schema_ref)
+        , m_schema_ref(*std::get<0>(m_schema))
```

Comment:
I think the copy constructor should perform a full copy, including the data.

Reply:
Done